### PR TITLE
Fixes #7766 - performance enhancements for DHCP on windows

### PR DESCRIPTION
**Pull request by briangann**
These changes enhances the DHCP smartproxy for windows (native_ms.rb) by providing sub-second responses for most queries. (This is as fast as you can get netsh.exe to respond)

Key changes:
- Constructs a script file to be run by netsh, and parses the output.
- Invokes netsh.exe with correct parameters
- Bugfix in the report function - would crash when no option exists

Benefit:
Large DHCP scopes take a significant amount of time to enumerate with the previous method of running netsh for each query. The smartproxy times out or takes up to 6 minutes on a /22 network, even on very fast dedicated hardware.

Performance:
Usage of this modification has shown it to perform at sub-second response even with large scopes.
A /22 scope previously would take over 6 minutes to enumerate and the browser would time out or appear to be "stuck."

### CODE REVIEW

**Comment by dmitri-d** [comment]
[test]

File: modules/dhcp/providers/server/native_ms.rb [hunk 1]
```diff
@@ -92,30 +93,50 @@ def loadSubnetData subnet
       cmd = "scope #{subnet.network} show reservedip"
       msg = "Enumerated hosts on #{subnet.network}"
 
-      # Extract the data
+      netsh_tmp = Tempfile.new('smartproxydhcp')
+      netsh_tmp.write("dhcp server\n")
+      all_entries = Hash.new()
+      # Extract the data and build a scriptfile
       execute(cmd, msg).each do |line|
```
Comment (dmitri-d) [inline modules/dhcp/providers/server/native_ms.rb:100]: could you move cmd and msg to the line above `execute(cmd, msg)`? I'd get rid of `msg` altogether.

File: modules/dhcp/providers/server/native_ms.rb [hunk 2]
```diff
@@ -92,30 +93,50 @@ def loadSubnetData subnet
...
-              # this is not a lease, rather reservation
-              # but we require option 12(hostname) to be defined for our leases
-              # workaround until #1172 is resolved.
-              Proxy::DHCP::Lease.new opts
-            end
-          rescue Exception => e
-            logger.warn "Skipped #{line} - #{e}"
-          end
+          opts = {:subnet => subnet, :ip => ip, :mac => mac}
+          netsh_tmp.write("scope #{opts[:subnet].network} Show ReservedOptionValue #{ip}\n")
```
Comment (dmitri-d) [inline modules/dhcp/providers/server/native_ms.rb:107]: This is going to be pretty hard to test. You can extract lines 96-111 into two methods: the first one will build all_entries hash, the second will generate the script file based on the all_entries. 

Can you add a test to verify script file generation please?

File: modules/dhcp/providers/server/native_ms.rb [hunk 3]
```diff
@@ -196,6 +218,61 @@ def execute cmd, msg=nil, error_only=false
...
+        std_in.close  unless std_in.nil?
+        std_out.close unless std_in.nil?
+        std_err.close unless std_in.nil?
+      end
+      report msg, response, error_only
+      response
+    end
+
+    # convert scriptfile output into a hash for fast queries
+    def parse_script_output_to_hash content
```
Comment (dmitri-d) [inline modules/dhcp/providers/server/native_ms.rb:247]: Can you add a test for this please?

File: modules/dhcp/providers/server/native_ms.rb [hunk 4]
```diff
@@ -228,21 +305,26 @@ def parse_options response
       options  = {}
       response.each do |line|
         line.chomp!
+        #logger.debug "PARSELINE :#{line}"
```
Comment (dmitri-d) [inline modules/dhcp/providers/server/native_ms.rb]: Did you forget to remove this?

File: modules/dhcp/providers/server/native_ms.rb [hunk 5]
```diff
@@ -92,30 +93,50 @@ def loadSubnetData subnet
...
+          logger.info "New reservation #{existing_options.inspect}"
+          Proxy::DHCP::Reservation.new existing_options.merge({:deleteable => true})
+        else
+          # this is not a lease, rather reservation
+          # but we require option 12(hostname) to be defined for our leases
+          # workaround until #1172 is resolved.
+          logger.info "New Lease #{existing_options.inspect}"
+          Proxy::DHCP::Lease.new existing_options
+         end
+      end
```
Comment (dmitri-d) [inline modules/dhcp/providers/server/native_ms.rb:137]: You can extract lines 120-138 into a methods, which will help with testing. Can you add a test for this please?

**Comment by dmitri-d** [comment]
There are also code style issues, you can see them if you follow http://ci.theforeman.org/job/test_proxy_develop_pull_request/693/.

**Comment by dmitri-d** [comment]
Please squash the commits, and see my and bot's comments.

**Comment by briangann** [comment]
I'm git-challenged sorry, I've tried squash/commit and it keeps kicking out the same errors. Any assistance would be appreciated.

**Comment by dmitri-d** [comment]
`git rebase -i HEAD~4` should do the trick. You'll be given a choice which commits to squash, which is all but one (the last one is my usual pick). You'll be prompted for a commit message after that, please change it to the format that the bot insists on.

You never responded to my comments re: tests -- any chance you can add some?

**Comment by briangann** [comment]
I will create the tests this weekend, and split off some of the code as suggested to make it more testable.  The squash finally worked, I think :)  Thanks! Brian

**Comment by dmitri-d** [comment]
@briangann: let me know if you need a hand with tests!

**Comment by dmitri-d** [comment]
closing in favour of https://github.com/theforeman/smart-proxy/pull/294. Refactored the code a little bit, added a couple of tests and fixed broken ones.

**Pull request closed without merging**